[Draft] Prediction on unseen data (and bugfixes)

davidwyld opened the pull request:
Copied from: https://github.com/RSE-Sheffield/deepquest/issues/53#issuecomment-600666616

The interface now follows the form:
```shell
dq predict --model foo/bar/epoch_x.h5 --dir pred_dir --file baz
```
with optional arguments `--dataset`, `--changes`, `--save_path`, `--evalset`
similarly we can do:
```python
dq.predict(model, dataset=None, directory=None, filename=None, save_path=None, evalset=None, changes={})
```
if no dataset file is specified (which is best I think), then the function `findDataset()` looks in the directory where the model is (the standard output location from training) for a file of the name `'Dataset*.pkl` and raises an assert error if anything fewer or more than 1 are found. (@drj11 do you have any thoughts on whether I should use a different error type here?) Alternatively, we can specify a file.

**Assumptions**
Following the underlying assumptions of nmt-keras on how data is provided, this assumes that:
1. The data is provided as two (prepared) files of the sort `test.mt` and `test.src`
2. These two files are in the same directory.
3. They have the same basename, ie 'test'
4. Their file extensions are those provided in the parameters 'SRC_LAN' and 'TRG_LAN' such as 'src' and 'mt' or 'en' and 'de'.

*Note 1*: The `'NO_REF'` parameter was something that was causing confusion for me. During prediction, we take the params dictionary was from training, so that prediction is done with the same parameters. However, if `'NO_REF'=False` (i.e. there is a reference) then some of the processes in Keras Wrapper are looking for the reference data. (I can explain more if needed).
One simple solution is just to always set `'NO_REF'=True` during prediction. Can anyone see a problem with that?

*Note 2*: @fredblain if you want to do this with config files, that's fine, I'll just get this working first, then move over to that method when this is working as expected to save me any confusion.

*Note 3*: @fredblain, I haven't been able to reproduce your bug from above yet. I will look into this next. I remember that Keras Wrapper has some messy logic when it can't find an h5 file of the form it is expecting then it assumes that your model is in an h5+json format, then throws an error when it can't find that.

Commit:
add options in predict.main() to predict on unseen text files

add input params for predict.main() for the directory containing files to be predicted on
and for common filename for those files.
dQ expects files to have the form test.src or test.mt and for them to be in the same dir
parameters['TEXT_FILES'] and 'DATA_ROOT_PATH' are modified for prediction
so that they refer to the specified files.

Commit:
add options in predict.main() to predict on unseen text files

add input params for predict.main() for the directory containing files to be predicted on
and for common filename for those files.
dQ expects files to have the form test.src or test.mt and for them to be in the same dir
parameters['TEXT_FILES'] and 'DATA_ROOT_PATH' are modified for prediction
so that they refer to the specified files.

Commit:
add cli options for prediction files

Commit:
add cli options for prediction files

Commit:
clarify how changes are specified in cli

Commit:
modifications to allow predicting on arbitrary files

Commit:
modifications to allow predicting on arbitrary files

Commit:
make dataset an optional parameter

If no dataset is specified as an  input to predict, the function findDataset() will
look for one of the form 'Dataset*.pkl' in the model directory.
If number of datasets found is not exactly one, it throws an assert error.

Commit:
make dataset an optional parameter

If no dataset is specified as an  input to predict, the function findDataset() will
look for one of the form 'Dataset*.pkl' in the model directory.
If number of datasets found is not exactly one, it throws an assert error.

Commit:
update tests

Commit:
update tests

Commit:
fix filepath bug

Commit:
fix filepath bug

drj11 commented on deepquest/predict.py:
```diff
@@ -210,25 +210,47 @@ def check_params(params):
+    for file in os.listdir(os.path.split(model)[0]):
+        if fnmatch.fnmatch(file,'Dataset*.pkl'):
+                datasets.append(file)
+    assert len(datasets)==1, "Expected to find exactly one dataset in {}/ but found {}\n{}\nPlease specify a dataset file.".format(model, len(datasets), datasets)
```
`assert` statements get removed if you "compile" Python with `-O`. For that reason, for things that you actually want to get checked, they are better as an `if :` with a `raise` statement:

    if len(datasets) != 1:
        raise Exception("Expected to find exactly one dataset in {}/ but found {}\n{}\nPlease specify a dataset file.".format(model, len(datasets), datasets))

My suggested code here uses the builtin `Exception` class which is the superclass of all ordinary exceptions. You are welcome to use another one, maybe `OSError`, or create your own: `class DatasetError(Exception): pass`. But really it's probably not worth worrying about too much.

fredblain replied:
This block has been removed in the new version of `predict()`.
Mark as resolved.

fredblain committed:
removed unused function.

fredblain committed:
Merge branch 'hotfix-predict' of github.com:RSE-Sheffield/nmt-keras into hotfix-predict

fredblain committed:
rewriting predict.py

predict.py has been revised to predict on unseen data
It loads a config file from which it obtain i) the path to and config of a pretrained QE model;
ii) source and target input files for which we want to predict estimates;
iii) build a Dataset object from those input files;
iv) maps the input IDs of the model and the new Dataset;
v) produce the predictions and post-process them
vi) dump the predictions into a 'predictions.txt' file (path given by config).

fredblain committed:
update __main__.py to have same API between train and predict.

fredblain committed:
update __init__.py to have same API between train and predict.

fredblain committed:
updating setparameters() to consider pretrained config over default one.

fredblain commented:
The new version of `predict()` comes with a series of \#TODO we need to address.

Commit:
remove redundant predict callback

this callback was added in error and is unused

davidwyld commented:
Hi @fredblain, so we have these two problems with prediction at present, right?
1. Prediction for BERT tokenized models, documented here: https://github.com/RSE-Sheffield/deepquest/issues/53
2. and word-level prediction, which I find a little confusing, so please forgive my ignorance:

For word-level prediction, line 50 of `predict.py`:  `raw_predictions = model.predictNet(dataset, predict_parameters)` returns a shape (500, 100, 5) array.
Am I right in thinking that we want to do something similar to what is done in `eval_word_qe()` when there are no gold-labels:
```python
        for p in thresholds:
            # predictions from the wordQE model
            y_pred = []
            for i in range(len(target_text)):
                # we use the length of the target sent to retrieve the number
                # of labels that we need to produce, otherwise padding to 70...
                mt_sent = target_text[i].split(' ')
                # logging.info('y_init[i]: %s' % y_init[i])
                for j in range(len(mt_sent)):
                    pred_word = y_init[i][j]
                    if pred_word[vocab['words2idx']['OK']] >= p:
                        y_pred.append('OK')
                    else:
                        y_pred.append('BAD')
            # logging.info('y_pred: %s' % y_pred)
            y_pred = np.array(y_pred)
            res_list[p] = y_pred

        return {'pred_categorical': res_list}
```

fredblain committed:
removing (temporarily) the option to tune emebdding layers (which prevents reloading the model, and causes the model to be huge on disk).

fredblain committed:
quick fix to predict on unseen data with word-level (default) BiRNN QE models.

fredblain committed:
quick fix to predict on unseen data with word-level (default) BiRNN QE models.

(RSE-Sheffield/deepquest/issues/53)

fredblain committed:
Merge branch 'hotfix-predict' of github.com:RSE-Sheffield/nmt-keras into hotfix-predict

fredblain reviewed (approved):
This PR is very dirty but I have no choice to approve it in order to move forwards.

fredblain committed:
Merge branch 'deepquest-dev' into hotfix-predict
